Port GF.jl to Keldysh.jl 0.6.0

krivenko opened the pull request:
This PR is still work-in-progress and contains a few issues to be resolved.

- [ ] There are now 9 methods of `computegf()` instead of 3. The extra factor 3 comes from the time grid types (`FullTimeGrid`, `KeldyshTimeGrid` and `ImaginaryTimeGrid`). Is there a way to treat them all simultaneously (bodies of the methods are nearly identical) without compromising type stability?
- [x] Parallelization of `computegf()` does not work yet. Before I was using a `SharedArray`, which was concurrently filled by a `@distributed for` loop. At the end of the calculation I simply copied contents of the array into a new `TimeGF` object. Now that Keldysh.jl has a variety of storage schemes, it should provide some tools to implement parallelism/scatter-gather functionality, I believe.
- [x] Is there a way to set a single matrix element of a (t1, t2) slice of a matrix-valued GF? Right now I have to write
```
val_mat = zeros(T, norb, norb)
val_mat[c_n, cdag_n] = (greater ? -1im : 1im) * val
gf[t1, t2] += val_mat
```
- [x] Unit test: What is the best way to compare an `ImaginaryTimeGF`/`TimeInvariantKeldyshTimeGF` object with the respective portion of another `TimeInvariantFullTimeGF` object?

krivenko committed:
Port GF.jl to Keldysh.jl 0.6.0

krivenko committed:
travis: Install Keldysh.jl v0.6.0 (not yet released)

krivenko committed:
test: Finish the GF unit test

krivenko committed:
Restore parallelization of computegf()

krivenko committed:
Adjust code to final release of Keldysh.jl v0.6.0

krivenko committed:
Require Keldysh.jl 0.6.1

kleinhenz commented on README.md:
```diff
@@ -24,7 +24,7 @@ typing the following commands in Julia's `Pkg` REPL:
     add HDF5
     add ArgParse
-    add https://github.com/kleinhenz/Keldysh.jl.git#v0.5.1
+    add https://github.com/kleinhenz/Keldysh.jl.git#v0.6.0
```
Now v0.6.1

kleinhenz commented on src/GF.jl:
```diff
@@ -18,54 +18,43 @@
+  )
+
+  @sync @distributed for job = 1:njobs
+    (c_n, c_index), (cdag_n, cdag_index), (t1, t2) = all_jobs[job]
```
can we factor this inner code to compute a single element of the Green's function out into its own function?

kleinhenz commented on src/GF.jl:
```diff
@@ -18,54 +18,43 @@
-    end
-    β = length(im_b)
-  end
+function _computegf!(ed::EDCore,
```
thinking about this a little more, I think serial should probably be the default since the function is not that slow and there are a few different ways you could think about parallelizing it (threads, processes etc). If the inner part is factored out into a separate function then it wouldn't be too hard to write different parallel versions.

krivenko replied:
Okay, what do you think about the following solution, https://github.com/krivenko/KeldyshED.jl/pull/9/commits/7d31c4fef01baaeed7015ffe4df37ef268b372aa ?

krivenko committed:
README: Depend on Keldysh.jl v0.6.1

krivenko committed:
[GF] Add two new methods of computegf()

* The added methods compute a single GF element at times t1, t2 for
  given indices of creation/annihilation operators. One of the methods
  takes a precomputed list of energy levels and a density matrix so that
  recomputing them can be avoided in repeated calls.

* The existing methods of computegf() have been extended to take an
  extra functor (<:AbstractGFFiller) which controls how a GF object is
  built out of its individual elements. There are two implementations of
  AbstractGFFiller available, SerialGFFiller (uses a simple for loop to
  fill the GF container, the default) and DistributedGFFiller
  (uses a @distributed for).

* Unit test has been updated.

krivenko committed:
[GF] Make gf_filler a keyword argument

krivenko committed:
[GF] Make `en` and `ρ` arguments of `computegf()` keyworded

krivenko committed:
Bump version to 0.6.0
